docs/proposals/20191016-clusterctl-redesign.md

```diff
@@ -0,0 +1,271 @@
+$ kubectl apply --kubeconfig=my-first-cluster.kubeconfig -f MY_CNI
+```
+
+The internal flow for the above commands is the following:
+
+**clusterctl init (AWS):**
+
+![components](images/clusterctl-redesign/init.png)
+
+Please note that “day one” experience:
+
+1. Is based on a list of pre-configured provider repositories; it is possible to add new provider configurations to the list of known providers by using the clusterctl config file (see day 2 operations for more details about this).
+2. Assumes init will replace env variables defined in the components yaml read from provider repositories, or error out if such variables are missing.
+3. Assumes providers will be installed in the namespace defined in the components yaml; at the time of this writing, all the majoir providers ships with components yaml creating dedicated namespaces. This could be customized by specifying the --target-namespace flag.
+4. Assumes to not change the namespace each provider is watching on objects; at the time of this writing, all the majoir providers ships with components yaml watching for objects in all namespaces. Pending [1490](https://github.com/kubernetes-sigs/cluster-api/issues/1490) issue addressed, this could be customized by specifying the --watching-namespace flag.
+
+**clusterctl config:**
+
+![components](images/clusterctl-redesign/config.png)
+
+Please note that “day one” experience:
+
+1. Assumes only one infrastructure provider and only one bootstrap provider installed in the cluster; in case of more than one provider installed, it will be possible to specify the target provider using the --infrastructure and --bootstrap flag.
+2. In order to fetch template from a provider repository, Assumes a naming convention should be established (details TBD).
+3. Similarly, also template variables should be defined (details TBD).
+
+#### Day 2 operations
+
+The clusterctl command will provide support for following operations
+
+- Add more provider configurations using the clusterctl config file [1]
+- Override default provider configurations using the clusterctl config file [1]
+- Add more providers after init, eventually specifying the target namespace and the watching namespace [2]
+- Add more instances of an existing provider in another namespace/with a non-overlapping watching namespace [2]
+- Get the yaml for installing a provider, thus allowing full customization for advanced users
+- Create additional clusters, eventually specifying template parameters
+- Get the yaml for creating a new cluster, thus allowing full customization for advanced users
+- Pivot the management cluster to another management cluster (see next paragraph)
+- Upgrade a provider (see next paragraph)
+- Delete a provider (see next paragraph)
+- Reset the management cluster (see next paragraph)
+
+*[1] clusterctl will provide support pluggable provider repository implementations. Current plan is to support:*
```

Comment:
I see my previous comment regarding not using repositories for the provider yamls is partially covered as local file system is also supported. However, I still think external repositories should be handled outside the library (e.g cli)

Reply:
I don't understand what you mean. Could you please give a specific example of what you dislike, and how you would change it?

Reply:
Simply a separation of concern. As described in the components, there are three layers: cli, client and lib. IMHO the lib should use the simplest non-opinionated source (local files) and let the upper layers (i.e. cli or external tools) where to take this files from.

Reply:
This sounds like an implementation detail. I would guess the code takes in something like an `io.ReadCloser`. @fabriziopandini would you agree? @pablochacin ok to mark this resolved?

Reply:
Yes, this is definitely an implementation detail. It was triggered because in the section above it says " clusterctl internal pkg/lib is expected to access provider repositories" which is also an implementation detail.

Reply:
The upper-level layer will be allowed to inject their repository implementation (pluggable provider repository implementations).

But we should consider that most probably not all the library consumers want to implement a custom repository, so the internal lib is going to provide the default repository implementations listed in the doc, not the only file system.

Reply:
@pablochacin just checking in - how are you feeling about this question?